source/common/router/router.cc

```diff
@@ -191,6 +191,17 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   timeout_ = FilterUtility::finalTimeout(*route_entry_, headers);
+
+  // If this header is set with any value, use an alternate response code on timeout
```

Comment:
I don't see any downside. The user has to configure it, and we only allow this feature on internal calls.